README
   _     _                                     _   _          
  | |   (_)_ __   __ _ _   _  __ _   _ _      | | / \     _ _ 
  | |   | | '_ \ / _` | | | |/ _` | (_|_)  _  | |/ _ \   (_|_)
  | |___| | | | | (_| | |_| | (_| |  _ _  | |_| / ___ \   _ _ 
  |_____|_|_| |_|\__, |\__,_|\__,_| (_|_)  \___/_/   \_\ (_|_)
                 |___/                                        
   _____ _           _ ____        _            
  |  ___(_)_ __   __| |  _ \  __ _| |_ ___  ___ 
  | |_  | | '_ \ / _` | | | |/ _` | __/ _ \/ __|
  |  _| | | | | | (_| | |_| | (_| | ||  __/\__ \
  |_|   |_|_| |_|\__,_|____/ \__,_|\__\___||___/
                                                


This is the README for Lingua::JA::FindDates version 0.020.

Lingua::JA::FindDates is a "module" for the Perl computer programming language,
a library of computer code to install on a computer.  This document contains
four sections:

1. About - what the module does

2. Documentation - how to learn more about the module

3. Installation - how to install this module on a computer

4. Help - what to do if you get stuck

-----------------------------------------------------------------------------

1. ABOUT

Lingua::JA::FindDates - scan text to find dates in a Japanese format

This module offers pattern matching of dates in the Japanese
language. Its main routine, "subsjdate" scans a text and finds
things which appear to be Japanese dates.

The module recognizes the typical format of dates with the year first,
followed by the month, then the day, such as 平成20年七月十日
"(Heisei nijūnen shichigatsu tōka)". It also recognizes combinations
such as years alone, years and months, a month and day without a year,
fiscal years (年度, "nendo"), parts of the month, like 中旬 (chūjun,
the middle of the month), and periods between two dates.

It recognizes both the Japanese-style era-base year format, such as 
"平成２４年" (Heisei) for the current era, and European-style Christian
era year format, such as 2012年. It recognizes several forms of
numerals, including the ordinary ASCII numerals, 1, 2, 3; the "wide"
or "double width" numerals sometimes used in Japan, １, ２, ３; and
the kanji-based numeral system, 一, 二, 三. It recognizes some special
date formats such as 元年 for the first year of an era. It recognizes
era names identified by their initial letters, such as S41年 for Shōwa
41 (1966). It recognizes dates regardless of any spacing which might
be inserted between individual Japanese characters, such as 
"平 成 二 十 年 八 月".

The input text must be marked as Unicode, in other words character
data, not byte data.

The module has been tested on several hundreds of documents, and it
should cope with all common Japanese dates. If you find that it cannot
identify some kind of date within Japanese text, please report that as
a bug.

-----------------------------------------------------------------------------

2. DOCUMENTATION

You can read the documentation for the module online at the following websites:

    * http://search.cpan.org/perldoc?Lingua::JA::FindDates
    * http://metacpan.org/release/Lingua-JA-FindDates

(These links go to the latest version of the module.)

After installing the module, you can read the documentation on your computer
using

    perldoc Lingua::JA::FindDates

-----------------------------------------------------------------------------

3. INSTALLATION

This module requires Perl version 5.1.1 or later.

To install the module from CPAN, use

    cpan Lingua::JA::FindDates

If you have the App::cpanminus installer, you may prefer

    cpanm Lingua::JA::FindDates

To install the module from the source file, Lingua-JA-FindDates-0.020.tar.gz,
follow this sequence of commands:

    tar xfz Lingua-JA-FindDates-0.020.tar.gz
    cd Lingua-JA-FindDates-0.020
    perl Makefile.PL
    make
    make install

If you want to test the module before installing it, use "make test" after
"make" and before "make install".

-----------------------------------------------------------------------------

4. HELP

To get help with the module, you can email the author, Ben Bullock, at
<bkb@cpan.org>. If you think there is a problem in the module, you can report a
bug at 

<https://github.com/benkasminbullock/Lingua--JA--FindDates/issues>,

or if you want to alter the source code of Lingua::JA::FindDates, try the
public repository on github at 

<https://github.com/benkasminbullock/Lingua--JA--FindDates>. 

-----------------------------------------------------------------------------

This README was written on Wed Feb  5 23:43:06 2014.

-----------------------------------------------------------------------------


Changes
0.020 2014-02-05

* Add README and MIN_PERL_VERSION in Makefile.PL.

0.019 2013-11-03

* Add squared symbols like ㍻ to the list of possibles.

0.018 2013-02-08

* Fix bug with Meiji start year (was 1870, now 1868).

0.017 2013-02-08

* Add two functions, "seireki_to_nengo" and "nengo_to_seireki".

0.016 2013-02-07

* Fix for tricky bug involving romaji.
* Code cleanup.

0.015 2012-07-27

* Move documentation to lib/Lingua/JA/FindDates.pod.

0.014 2012-07-27

* Add bugtracker, repository.

0.013 2012-03-19

* Add "subsjdate" script.

0.012 2012-03-19

* Documentation improvements.
* Change to ExtUtils::MakeMaker.
* Version control for source code.

0.011 2010-09-18

* Add meta files
* Handle apostrophe dates

0.010 2009-06-10

* Match only horizontal whitespace to prevent mismatches

0.009 2008-09-28

* Bug fix with "jun" disappearing sometimes.

0.008 2008-08-17

* Documentation fixes (see www.annocpan.org for details).
* Bug fixes (see rt.cpan.org for details).

0.007 2008-07-22

* Rewrote some of the documentation
* Added nyoro 0xFF5E ("Fullwidth tilde") for Microsoft Word nyoro.
* Changed to Module::Install.
* Stopped exporting kanji2number.

0.006 2008-07-17

* Fixed a bug with something like the second 日 in １月１日 日本人論
  being taken for Sunday.
* Added 元年 to regex
* Added note on bug 56902 in Perl.
* Changed interfaces for callbacks and for passing date information.
* Added intervals with weekdays like 平成９年１０月１７日（火）～２０日（金）
* Trimmed documentation.

0.005 2008-07-11

* Added support for ten day periods like 中旬 etc. 
* Added support for intervals like 平成１０年３月１～４日.

0.004 2008-07-08

* Fixed bugs with fiscal years like ２０００年度, turned off
  verboseness and added test to ensure that $verbose is switched off
  for release versions to avoid this happening again in the future.

0.003 2008-07-06

* Re-licensed to Perl licence. 

0.002 2008-07-04

* Fix mojibake in search.cpan.org by adding =encoding line at the top
  of the module.

0.001 2008-07-04

   - original version; created by h2xs 1.23 with options
       -X -n Lingua::JA::FindDates


example/test.pl
#! perl
use warnings;
use strict;
use utf8;
use Test::More tests => 39;
#use Test::More tests => 1;
use lib './lib/';
BEGIN { use_ok('Lingua::JA::FindDates') };

use Lingua::JA::FindDates qw/kanji2number subsjdate/;

$Lingua::JA::FindDates::verbose = 0;

# If you are on Cygwin on Japanese Windows, use the following:

binmode STDOUT,":encoding(cp932)";
binmode STDERR,":encoding(cp932)";
#binmode STDOUT,":utf8";
#binmode STDERR,":utf8";
#goto juntest;
#print '百百三十五', "\n";
#print JaDates::kanji2number ('3百三十五');
ok (kanji2number ('3百三十五') == 0, 'bad kanji number failure test');
ok (kanji2number ('二百三十五') == 235, 'kanji number');
ok (kanji2number ('二三五') == 235, 'kanji number');
ok (kanji2number ('二三五五') == 2355, 'kanji number');

my @tests= qw/平成２０年７月３日（木） H二十年七月三日(木曜日) 二千八年7月三日(木曜)/;
for my $d (@tests) {
#    print subsjdate ($d);
    ok (subsjdate ($d) eq 'Thursday, July 3, 2008', 
	'year + month + day + weekday');
}

sub mymakedate
{
        my ($year, $month, $date, $wday) = @_;
	return qw{Bad Mo Tu We Th Fr Sa Su}[$wday]." $year/$month/$date";
}

for my $d (@tests) {
#    print subsjdate ($d);
    ok (subsjdate ($d, undef, undef, \&mymakedate)
	eq 'Th 2008/7/3', 'makedate_callback');
}

my @tests2= ('昭和 ４１年　３月１６日', 'Ｓ４１年三月十六日', '千九百六十六年3月16日');
for my $c (@tests2) {
#    print "$d, [>>",subsjdate ($d),"<<]\n";
    ok (subsjdate ($c) eq 'March 16, 1966', 'year + month + day combination');
}
my @tests3= ('昭和 ４１年', 'Ｓ４１年', '千九百六十六年');
for my $c (@tests3) {
#    print "$c, [>>",subsjdate ($c),"<<]\n";
    ok (subsjdate ($c) eq '1966', 'year combination');
}
my @tests4= ('３ 月 １６日', '三月十六日', '3月16日');
for my $c (@tests4) {
#    print "$d, [>>",subsjdate ($d),"<<]\n";
    ok (subsjdate ($c) eq 'March 16', 'month + day combination');
}

my $test5 =<<EOF;
2008年07月03日 01:39:21 投稿
【西村修平・桜井誠編】平成20年7月2日毎日新聞社前抗議行動！
最低賃金法が大きく改正され、平成２０年７月１日から施行されました。
平成二十年七月四日
Opera 9.51(Windows版)を入れてみたけれども相變らず使ひ物にならない。何なんだらう。
日本義塾 平成二十年七月公開講義
公開日時：2008年06月28日 01時09分
更新日時：2008年06月29日 00時08分

◎【日本義塾七月公開講義】

　◎日　時　平成二十年七月二十五日（金）
　　　　　　午後六時半～九時（六時開場）
H20年7月壁紙
7月壁紙（1024×768）を作成しました（クリックすると拡大します）。
昭和４９年度
1999年度
EOF

my %jdates = 
(
'2008年07月03日' =>'July 3, 2008',
'平成20年7月2日' =>'July 2, 2008',
'平成２０年７月１日' =>'July 1, 2008',
'平成二十年七月四日' =>'July 4, 2008',
'平成二十年七月' =>'July 2008',
'2008年06月28日' =>'June 28, 2008',
'2008年06月29日' =>'June 29, 2008',
'平成二十年七月二十五日（金）' =>'Friday, July 25, 2008',
'H20年7月' => 'July 2008',
'7月' => 'July',
'七月' => 'July',
'昭和４９年度' => 'fiscal 1974',
'1999年度' => 'fiscal 1999',
);

sub replace_callback
{
    my ($data, $jdate, $edate) = @_;
#    print "[$jdate] [$edate] [$jdates{$jdate}]\n";
    ok ($jdates{$jdate} eq $edate, "replace_callback");
}

subsjdate ($test5, \&replace_callback);

#ok ($Lingua::JA::FindDates::verbose == 0, 'verbose option switched off by default');
my @tests_jun = ('昭和41年3月初旬', 'Ｓ４１年三月上旬');#, '千九百六十六年３月初旬');
for my $c (@tests_jun) {
#    print "$c, [>>",subsjdate ($c),"<<]\n";
#    print "Looking for $c\n";
#    print $c, " ", subsjdate($c),"\n";
    ok (subsjdate ($c) eq 'early March 1966', 'year/month/jun combination');
#ok(1);
}

my @tests_interval = 
('昭和41年3月1〜12日',
 '昭和41年3月1日〜12日',
 '昭和41年1〜12月',
 '昭和41年1月〜12月',);

for my $c (@tests_interval[0..1]) {
#    print "Looking for $c\n";
#    print $c, " ", subsjdate($c),"\n";
    ok (subsjdate($c) eq 'March 1-12, 1966', "two days interval");
}
for my $c (@tests_interval[2..3]) {
#    print "Looking for $c\n";
#    print $c, " ", subsjdate($c),"\n";
    ok (subsjdate($c) eq 'January-December 1966', "month-month interval");
}


lib/Lingua/JA/FindDates.pm
package Lingua::JA::FindDates;

use 5.010000;
require Exporter;
use AutoLoader qw(AUTOLOAD);
our @ISA = qw(Exporter);
@EXPORT_OK= qw/subsjdate kanji2number seireki_to_nengo nengo_to_seireki/;
our $VERSION = '0.020';
use warnings;
use strict;
use Carp;
use utf8;

# Kanji number conversion table.

my %kanjinums = 
(
    〇 => 0,
    一 => 1,
    二 => 2,
    三 => 3,
    四 => 4,
    五 => 5,
    六 => 6,
    七 => 7,
    八 => 8,
    九 => 9,
    十 => 10,
    百 => 100,
    # Dates shouldn't get any bigger than the following times a digit.
    千 => 1000,
); 

# The kanji digits.

my $kanjidigits = join ('', keys %kanjinums);

sub kanji2number
{
    my ($knum) = @_;
    return 1 if $knum eq '元';
    my @kanjis = split '', $knum;
    my $value = 0;
    my $keta = 1;
    my @values;
    while (1) {
	my $k = pop @kanjis;
	return $value if ! defined $k;
	my $val = $kanjinums{$k};
        # Make sure this kanji number is one we know how to handle.
	if (! defined $val) {
	    warn "can't cope with '$k' of input '$knum'";
	    return 0;
	}
        # If the value of the individual kanji is more than 10.
	if ($val >= 10) {
	    $keta = $val;
	    my $knext = pop @kanjis;
	    if (!$knext) {
		return $value + $val;
	    }
	    my $val_next = $kanjinums{$knext};
	    if (!defined $val_next) {
		warn "can't cope with '$knext' of input '$knum'.\n";
		return 0;
	    }
	    if ($val_next > 10) {
		push @kanjis, $knext;
		$value += $val;
	    }
            else {
		$value += $val_next * $val;
	    }
	}
        else {
            # $k is a kanji digit from 0 to 9, and $val is its value.
	    $value += $val * $keta;
	    $keta *= 10;
	}
    }
}

#  ____                               
# |  _ \ ___  __ _  _____  _____  ___ 
# | |_) / _ \/ _` |/ _ \ \/ / _ \/ __|
# |  _ <  __/ (_| |  __/>  <  __/\__ \
# |_| \_\___|\__, |\___/_/\_\___||___/
#            |___/                    

# Map "double-byte" or "double-width" numbers to single byte numbers
# (the usual ASCII numbers).

my $nums = '０１２３４５６７８９';
my @wnums = split '', $nums;
my %wtonarrow;
for (0..9) {
    $wtonarrow{$wnums[$_]} = $_;
}

my $jdigit = qr/[０-９0-9]/;

# A regular expression to match Japanese numbers

my $jnumber = qr/($jdigit+|[$kanjidigits]+)/x;

# A regular expression to match a Western year

my $wyear = qr/
                  (
                      $jdigit{4}
                  |
                      [$kanjidigits]?千[$kanjidigits]*
                  |
                      [\']$jdigit{2}
                  )
                  \s*年
              /x;

my $alpha_era = qr/
                      # If the H, S, T, or M is part of a longer
                      # string of romaji, do not match it.
                      (?<![A-ZＡ-Ｚ])
                      (?:
                          [H|Ｈ|S|Ｓ|T|Ｔ|M|Ｍ]
                      )
                  /x;

# The recent era names (Heisei, Showa, Taisho, Meiji). These eras are
# sometimes written using the letters H, S, T, and M.

my $jera = qr/($alpha_era|平成|昭和|大正|明治|㍻|㍼|㍽|㍾)/;

# A map of Japanese eras to Western dates. These are the starting year
# of the period minus one, to allow for that the first year is "heisei
# one" rather than "heisei zero".

my %jera2w = (
    H    => 1988,
    Ｈ   => 1988,
    平成 => 1988,
    '㍻' => 1988,
    S    => 1925,
    Ｓ   => 1925,
    昭和 => 1925,
    '㍼' => 1925,
    T    => 1911,
    Ｔ   => 1911,
    大正 => 1911,
    '㍽' => 1911,
    M    => 1867,
    Ｍ   => 1867,
    明治 => 1867,
    '㍾' => 1867,
);

# Japanese year, with era like "Heisei" at the beginning.

my $jyear = qr/
                  $jera
                  \h*
                  # Only match up to one or two of these digits, to
                  # prevent unlikely matches.
                  (
                      $jdigit{1,2}
                  |
                      [$kanjidigits]{1,2}
                  |
                      # The first year of an era, something like
                      # "昭和元年" (1926, the first year of the Showa era).
                      元 
                  )
                  \h*
                  年
              /x;

# The "jun" or approximately ten day periods (thirds of a month)

my %jun = qw/初 1 上 1 中 2 下 3/;

# The translations of the "jun" above into English.

my @jun2english = (
    'invalid',
    'early ',
    'mid-',
    'late ',
);

# Japanese days of the week, from Monday to Sunday.

my $weekdays = '月火水木金土日';
my @weekdays = split '',$weekdays;

# Match a string for a weekday, like 月曜日 or (日)
# The long part (?=\W) is to stop it from accidentally matching a
# kanji which is part of a different word, like the following:
#平成二十年七月一日
#    日本論・日本人論は非常に面白いものだ。

my $match_weekday =
    qr/[\（(]?
       ([$weekdays])
       (?:(?:(?:曜日|曜)[)\）])|[)\）]|(?=\W))
      /x;

# my $match_weekday = '[（(]?(['.$weekdays.'])(?:曜日|曜)?[)）]?';

# Match a day of the month, like 10日

my $match_dom = qr/$jnumber\h*日/;

# Match a month

my $match_month = qr/$jnumber\h*月/;

# Match a "jun" (a third of a month).

my $jun_keys = join ('', keys %jun);

my $match_jun = qr/([$jun_keys])\h*旬/;

# Match a month+jun

my $match_month_jun = qr/$match_month\h*$match_jun/;

# Match a month and day of month pair

my $match_month_day = qr/$match_month\h*$match_dom/;

# Match a Japanese year, month, day string

my $matchymd = qr/
                     $jyear
                     \h*
                     $match_month_day
                 /x;

# Match a Western year, month, day string

my $matchwymd = qr/$wyear\h*$match_month_day/;

# Match a Japanese year and month only

my $match_jyear_month = qr/$jyear\h*$match_month/;

# Match a Western year and month only

my $match_wyear_month = qr/$wyear\h*$match_month/;

# Match a month, day, weekday.

my $match_month_day_weekday = qr/$match_month_day\h*$match_weekday/;

# Separators used in date strings
# Microsoft Word uses Unicode 0xFF5E, the "fullwidth tilde", for nyoro symbol.

my $separators = qr/\h*[〜−~]\h*/;
 
# =head2 Matching patterns

# I<The module can be used without reading this section>.

# The Japanese date regular expressions are stored in an array
# B<jdatere> containing a pair of a regular expression to match a kind
# of date, and a string like "ymdw" which contains letters saying what
# to do with $1, $2, etc. from the regular expression. The array
# B<jdatere> is ordered from longest match (like "year / month / day /
# weekday") to shortest (like "year" only). For example, if the first
# letter is "y", then $1 is a year in Western format like 2008, or if
# the third letter is "w", then $3 is the day of the week, from 1 to 7.

# =over

# =item e

# Japanese era (string).

# =item j

# Japanese year (string representing small number)

# =item x

# empty month and day

# =item m

# month number (from 1 to 12, 13 for a blank month, 0 for an invalid month)

# =item d

# day of month (from 1 to 31, 32 for a blank day, 0 for an invalid day)

# =item w

# weekday (from Monday = 1 to Sunday = 7, zero or undefined for an
# invalid weekday)

# =item z

# jun (旬), a ten day period.

# =item 1

# After another code, indicates the first of a pair of two things. For
# example, the matching code for

#   平成９年１０月１７日〜２０日

# is

#   ejmd1d2

# =back

# =cut

#  _     _     _            __                                     
# | |   (_)___| |_    ___  / _|  _ __ ___  __ _  _____  _____  ___ 
# | |   | / __| __|  / _ \| |_  | '__/ _ \/ _` |/ _ \ \/ / _ \/ __|
# | |___| \__ \ |_  | (_) |  _| | | |  __/ (_| |  __/>  <  __/\__ \
# |_____|_|___/\__|  \___/|_|   |_|  \___|\__, |\___/_/\_\___||___/
#                                         |___/                    

# This a list of date regular expressions.

my @jdatere = (

    # Match an empty string like 平成 月 日 as found on a form etc.

    [qr/
           $jyear
           (\h+)
           月
           \h+
           日
       /x,
     "ejx"],

    # Add match for dummy strings here!

    # Match a Japanese era, year, 2 x (month day weekday) combination

    [qr/
           $matchymd
           \h*$match_weekday
           $separators
           $match_month_day_weekday
       /x,
     "ejm1d1w1m2d2w2"],

    # Match a Japanese era, year, month 2 x (day, weekday) combination

    [qr/
           $matchymd
           $match_weekday
           $separators
           $match_dom
           \h*
           $match_weekday
       /x, 
     "ejmd1w1d2w2"],

    # Match a Japanese era, year, month 2 x day combination

    [qr/
           $matchymd
           $separators
           $match_dom
           \h*
           $match_weekday
       /x,
     "ejmd1d2"],

    # Match a Western year, 2x(month, day, weekday) combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
           $separators
           $match_month_day_weekday
       /x,
     "ym1d1w1m2d2w2"],

    # Match a Western year, month, 2x(day, weekday) combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
           $separators
           $match_dom
           \h*
           $match_weekday
       /x,
     "ymd1w1d2w2"],

    # Match a Western year, month, 2x(day) combination

    [qr/
           $matchwymd
           $separators
           $match_dom
       /x,
     "ymd1d2"],

    # Match a Japanese era, year, month1 day1 - month 2 day2 combination

    [qr/
           $matchymd
           $separators
           $match_month_day
       /x,
     "ejm1d1m2d2"],

    # Match a Japanese era, year, month1 - month 2 combination

    [qr/
           $jyear
           \h*
           $jnumber
           \h*月?
           $separators
           $match_month
       /x, "ejm1m2"],

    # Match a Japanese era, year, month, day1 - day2 combination

    [qr/
           $match_jyear_month
           \h*
           $jnumber
           \h*日?
           $separators
           $match_dom
       /x,
     "ejmd1d2"],

    # Match a Japanese era, year, month, day, weekday combination

    [qr/
           $matchymd
           \h*
           $match_weekday
       /x,
     "ejmdw"],

    # Match a Japanese era, year, month, day

    [qr/$matchymd/,
     "ejmd"],

    # Match a Japanese era, year, month, jun

    [qr/
           $match_jyear_month
           \h*
           $match_jun
       /x,
     "ejmz"],

    # Match a Western year, month, day, weekday combination

    [qr/
           $matchwymd
           \h*
           $match_weekday
       /x,
     "ymdw"],

    # Match a Western year, month, day combination

    [qr/$matchwymd/,
     "ymd"],

    # Match a Western year, month, jun combination

    [qr/
           $match_wyear_month
           \h*
           $match_jun
       /x,
     "ymz"],

    # Match a Japanese era, year, month

    [qr/
           $jyear
           \h*
           $jnumber
           \h*
           月
       /x,
     "ejm"],

    # Match a Western year, month

    [qr/$match_wyear_month/,
     "ym"],

    # Match 2 x (month, day, weekday)

    [qr/
           $match_month_day_weekday
           $separators
           $match_month_day_weekday
       /x, 
     "m1d1w1m2d2w2"],

    # Match month, 2 x (day, weekday)

    [qr/
           $match_month_day_weekday
           $separators
           $match_dom
           \h*
           $match_weekday
       /x,
     "md1w1d2w2"],

    # Match month, 2 x (day, weekday)

    [qr/
           $match_month_day
           $separators
           $match_dom
       /x,
     "md1d2"],

    # Match a month, day, weekday

    [qr/$match_month_day_weekday/,
     "mdw"],

    # Match a month, day

    [qr/$match_month_day/,
     "md"],

    # Match a fiscal year (年度, nendo in Japanese). These usually don't
    # have months combined with them, so there is nothing to match a
    # fiscal year with a month.

    [qr/
           $jyear
           度
       /x,
     "en"],

    # Match a fiscal year (年度, nendo in Japanese). These usually don't
    # have months combined with them, so there is nothing to match a
    # fiscal year with a month.

    [qr/
           $wyear
           度
       /x,
     "n"],

    # Match a Japanese era, year

    [qr/$jyear/,
     "ej"],

    # Match a Western year

    [qr/$wyear/,
     "y"],

    # Match a month with a jun

    [
        qr/
              $match_month
              \h*
              $match_jun
          /x,
        "mz"
    ],

    # Match a month

    [
        qr/$match_month/,
        "m"
    ],
);

my @months = qw/Invalid
                January
                February
                March
                April
                May
                June
                July
		August
                September
                October
                November
                December
                MM/;

my @days = qw/Invalid
              Monday
              Tuesday
              Wednesday
              Thursday
              Friday
              Saturday
              Sunday/;

# This is a translation table from the Japanese weekday names to the
# English ones.

my %j2eweekday;

@j2eweekday{@weekdays} = (1..7);

# This is the default routine for turning a Japanese date into a
# foreign-style one.

sub make_date
{
    my ($datehash) = @_;
    my ($year, $month, $date, $wday, $jun) = 
	@{$datehash}{qw/year month date wday jun/};
    if (!$year && !$month && !$date && !$jun) {
	carp "No valid inputs\n";
	return;
    }
    my $edate = '';
    $edate = $days[$wday].", " if $wday;
    if ($month) {
	$month = int ($month); # In case it is 07 etc.
	$edate .= $months[$month];
	if ($jun) {
	    $edate = $jun2english[$jun] . $edate;
	}
    }
    if ($date) {
	$edate .= " " if length ($edate);
	$date = int ($date); # In case it is 07 etc.
	$date = "DD" if $date == 32;
	if ($year) {
	    $edate .= "$date, $year";
	} else {
	    $edate .= "$date";
	}
    } elsif ($year) {
	$edate .= " " if length ($edate);
	$edate .= $year;
    }
    return $edate;
}

# This is the default routine for turning a date interval into a
# foreign-style one, which is then substituted into the text.

sub make_date_interval
{
    my ($date1, $date2) = @_;
    my $einterval = '';
    my $usecomma;
    # The case of an interval with different years doesn't need to be
    # considered, because each date in that case can be considered a
    # single date.

    if ($date2->{month}) {
	if (!$date1->{month}) {
	    carp "end month but no starting month";
	    return;
	}
    }
    if ($date1->{month}) {
	if ($date1->{wday} && $date2->{wday}) {
	    if (! $date1->{date} || ! $date2->{date}) {
		carp "malformed date has weekdays but not days of month";
		return;
	    }
	    $usecomma = 1;
	    $einterval = $days[$date1->{wday}]  . " " . $date1->{date} .
		         ($date2->{month} ? ' '.$months[int ($date1->{month})] : ''). '-' .
		         $days[$date2->{wday}]  . " " . $date2->{date} . " " .
			 ($date2->{month} ? $months[int ($date2->{month})] : $months[int ($date1->{month})]);
	}
        elsif ($date1->{date} && $date2->{date}) {
	    $usecomma = 1;
	    if ($date1->{wday} || $date2->{wday}) {
		carp "malformed date interval: ",
		    "has weekday for one date but not the other one.";
		return;
	    }
	    $einterval = $months[int ($date1->{month})] . ' ' .
		         $date1->{date} . '-' .
			 ($date2->{month} ? 
			  $months[int ($date2->{month})] . ' ' : '') .
		         $date2->{date};
	}
        else { # no dates or weekdays
	    if ($date1->{date} || $date2->{date}) {
		carp "malformed date interval: only one day of month";
		return;
	    }
	    if (!$date2->{month}) {
		carp "start month but no end month or date";
		return;
	    }
	    $einterval = $months[int($date1->{month})] . '-' . 
		         $months[int($date2->{month})] .
			 $einterval;
	}
    }
    else { # weekday - day / weekday - day case.
	if ($date1->{wday} && $date2->{wday}) {
	    if (! $date1->{date} || ! $date2->{date}) {
		carp "malformed date has weekdays but not days of month";
		return;
	    }
	    $einterval = $date1->{wday}  . " " . $date1->{date} . '-' .
		         $date2->{wday}  . " " . $date2->{date};
	}
    }
    $einterval .= ($usecomma ? ', ': ' ').$date1->{year} if $date1->{year};
    return $einterval;
}

# If you want to see what the module is doing, set
#
#   $Lingua::JA::FindDates::verbose = 1;
#
# This makes L<subsjdate> print out each regular expression and reports
# whether it matched, which looks like this:
#
#   Looking for y in ([０-９0-9]{4}|[十六七九五四千百二一八三]?千[十六七九五四千百二一八三]*)\h*年
#   Found '千九百六十六年': Arg 0: 1966 -> '1966'

our $verbose = 0;

sub subsjdate
{
    # $text is the text to substitute. It needs to be in Perl's
    # internal encoding.
    # $replace_callback is a routine to call back if we find valid dates.
    # $data is arbitrary data to pass to the callback routine.
    my ($text, $callbacks) = @_;
    # Save doing existence tests.
    if (! $callbacks) {
        $callbacks = {};
    }
    if (! $text) {
        return $text;
    }
    # Loop through all the possible regular expressions.
    for my $datere (@jdatere) {
	my $regex = $$datere[0];
	my @process = split (/(?=[a-z][12]?)/, $$datere[1]);
        if ($verbose) {
            print "Looking for ",$$datere[1]," in ",$regex,"\n";
        }
	while ($text =~ /($regex)/g) {
	    my $date1;
	    my $date2;
            # The matching string is in the following variable.
	    my $orig = $1;
	    my @matches = ($2,$3,$4,$5,$6,$7,$8,$9);
            if ($verbose) {
                print "Found '$orig': " if $verbose;
            }
	    for (0..$#matches) {
		my $arg = $matches[$_];

		last if !$arg;
		$arg =~ s/([０-９])/$wtonarrow{$1}/g;
		$arg =~ s/([$kanjidigits]+|元)/kanji2number($1)/ge;
                if ($verbose) {
                    print "Arg $_: $arg " if $verbose;
                }
		my $argdo = $process[$_];
		if ($argdo eq 'e') { # Era name in Japanese
		    $date1->{year} = $jera2w{$arg};
		}
                elsif ($argdo eq 'j') { # Japanese year
		    $date1->{year} += $arg;
		}
                elsif ($argdo eq 'y') {
		    $date1->{year} = $arg;
		}
                elsif ($argdo eq 'n') {
		    $date1->{year} += $arg;
		    $date1->{year} = "fiscal ".$date1->{year};
		}
                elsif ($argdo eq 'm' || $argdo eq 'm1') {
		    $date1->{month} = $arg;
		}
                elsif ($argdo eq 'd' || $argdo eq 'd1') {
		    $date1->{date} = $arg;
		}
                elsif ($argdo eq 'm2') {
		    $date2->{month} = $arg;
		}
                elsif ($argdo eq 'd2') {
		    $date2->{date} = $arg;
		}
                elsif ($argdo eq 'w' || $argdo eq 'w1') {
		    $date1->{wday} = $j2eweekday{$arg};
		}
                elsif ($argdo eq 'w2') {
		    $date2->{wday} = $j2eweekday{$arg};
		}
                elsif ($argdo eq 'z') {
		    $date1->{jun} = $jun{$arg};
		}
                elsif ($argdo eq 'x') {
                    if ($verbose) {
                        print "Dummy date '$orig'.\n" if $verbose;
                    }
		    $date1->{date}  = 32;
		    $date1->{month} = 13;
		}
	    }
	    my $edate;
	    if ($date2) {
                # Date interval
		if ($callbacks->{make_date_interval}) {
		    $edate = 
                    &{$callbacks->{make_date_interval}} ($callbacks->{data},
                                                         $orig,
                                                         $date1, $date2);
		}
                else {
		    $edate = make_date_interval ($date1, $date2);
		}
	    }
            else {
                # Single date
		if ($callbacks->{make_date}) {
		    $edate = &{$callbacks->{make_date}}($callbacks->{data},
                                                        $orig,
                                                        $date1);
		}
                else {
		    $edate = make_date ($date1);
		}
	    }
            if ($verbose) {
                print "-> '$edate'\n" if $verbose;
            }


	    $text =~ s/\Q$orig\E/$edate/g;
	    if ($callbacks &&
                $callbacks->{replace}) {
		&{$callbacks->{replace}}($callbacks->{data}, $orig, $edate);
	    }
	}
    }
    return $text;
}

sub nengo_to_seireki
{
    my ($text) = @_;
    my %data;
    $data{count} = 0;

    my $out_text = subsjdate (
        $text, {
            make_date => \& nengo_to_seireki_make_date, 
            data => \%data,
        }
    );
    $out_text =~ s/#REPLACEME(\d+)REPLACEME#/$data{$1}/g;
    return $out_text;
}

sub nengo_to_seireki_make_date
{
    my ($data, $original, $date) = @_;
    if ($date->{year}) {
        $original =~ s/.*年/$date->{year}年/;
        my $count = $data->{count};
        $data->{$count} = $original;
        $data->{count}++;
        return "#REPLACEME${count}REPLACEME#";
    }
    else {
        return $original;
    }
}

sub seireki_to_nengo
{
    my ($text) = @_;
    my %data;
    $data{count} = 0;

    my $out_text = subsjdate (
        $text, {
            make_date => \& seireki_to_nengo_make_date, 
            data => \%data,
        }
    );
    $out_text =~ s/#REPLACEME(\d+)REPLACEME#/$data{$1}/g;
    return $out_text;
}

sub seireki_to_nengo_make_date
{
    my ($data, $original, $date) = @_;
    my $year = $date->{year};
    my @eras = (
        ['平成', 1989, 1, 8],
        ['昭和', 1926, 12, 25],
        ['大正', 1912, 7, 30],
        ['明治', 1868, 1, 25],
    );
    if (defined $year) {
        for my $era (@eras) {
            my $ename = $era->[0];
            my $eyear = $era->[1];
            my $emonth = $era->[2];
            my $eday = $era->[3];
            my $month = $date->{month};
            my $date = $date->{date};

            # This is a flag which says whether to perform a
            # substitution of the year or not.

            my $subs;

            # If the year is greater than the era year, or if the year
            # is the same as the era year and we do not know the
            # month, just replace.

            if ($year > $eyear ||
                ($year == $eyear && ! defined ($month))) {
                $subs = 1;
            }

            # If the year is the same, and there is a month

            elsif ($year == $eyear && defined ($month)) {

                # If there is a day of the month, then only substitute
                # if the month is greater than the changeover month,
                # or the month is the same, and the day of the month
                # is greater than or equal to the changeover day of
                # the month.

                if (defined ($date)) {
                    if ($month > $emonth ||
                        ($month == $emonth && $date >= $eday)) {
                        $subs = 1;
                    }
                }

                # If we don't know the day of the month, substitute if
                # the month is greater than or equal to the changeover
                # month.

                elsif ($month >= $emonth) {
                    $subs = 1;
                }
            }
            if ($subs) {

                # Only substitute if we need to.

                if ($original !~ /$ename/) {

                    # The year counting starts from 1, so we add 1 to
                    # the difference.

                    my $hyear = $year - $eyear + 1;
                    $original =~ s/\d+年/$ename${hyear}年/;
                }

                # Don't replace again, stop the loop.

                last;
            }
        }
    }
    my $count = $data->{count};
    $data->{$count} = $original;
    $data->{count}++;

    # This is a tag for substituting with.

    return "#REPLACEME${count}REPLACEME#";
}

1;

__END__
